File: 0048-rotate-image/0048-rotate-image.cs
public class Solution {
    public void Rotate(int[][] matrix) {
        int n = matrix.Length;
        int l = 0, r = n-1, t = 0, b = n-1;
        while (l < r){
            for (int i = 0; i < r-l; i++ ){
                int topLeft = matrix[t][l+i];
                matrix[t][l+i] = matrix[b-i][l];
                matrix[b-i][l] = matrix[b][r-i];
                matrix[b][r-i] = matrix[t+i][r];
                matrix[t+i][r] = topLeft;
            }
            l++;
            r--;
            t++;
            b--;
        }
    }
}

File: 0048-rotate-image/0048-rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        int left = 0, top = 0;
        int right = matrix.length - 1, bottom = matrix.length - 1;

        while (left < right){
            for (int i = 0; i < right - left; i++){
                int topLeft = matrix[top][left+i];
                matrix[top][left+i] = matrix[bottom-i][left];
                matrix[bottom-i][left] = matrix[bottom][right-i];
                matrix[bottom][right-i] = matrix[top+i][right];
                matrix[top+i][right] = topLeft;
            }
            left++;
            top++;
            right--;
            bottom--;
        }
    }
}